edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.DiagnosticsComponent/MetricsWorker.cs

```diff
@@ -0,0 +1,125 @@
+            this.storage = Preconditions.CheckNotNull(storage, nameof(storage));
+            this.uploader = Preconditions.CheckNotNull(uploader, nameof(uploader));
+            this.systemTime = systemTime ?? SystemTime.Instance;
+        }
+
+        public void Start(TimeSpan scrapingInterval, TimeSpan uploadInterval)
+        {
+            this.scrape = new PeriodicTask(this.Scrape, scrapingInterval, scrapingInterval, Log, "Metrics Scrape");
+            this.upload = new PeriodicTask(this.Upload, uploadInterval, uploadInterval, Log, "Metrics Upload");
+        }
+
+        async Task Scrape(CancellationToken cancellationToken)
+        {
+            using (await this.scrapeUploadLock.LockAsync(cancellationToken))
+            {
+                Log.LogInformation("Scraping Metrics");
+                List<Metric> metricsToPersist = new List<Metric>();
+                int numScrapedMetrics = 0;
+                foreach (var scrapedMetric in await this.scraper.ScrapeEndpointAsync(cancellationToken))
+                {
+                    numScrapedMetrics++;
+                    // Get the previous scrape for this metric
+                    if (this.metrics.TryGetValue(scrapedMetric.GetMetricKey(), out Metric oldMetric))
+                    {
+                        // If the metric is unchanged, do nothing
+                        if (oldMetric.Value.Equals(scrapedMetric.Value))
+                        {
+                            continue;
+                        }
+
+                        // if the metric has changed, write the previous metric to disk
+                        metricsToPersist.Add(oldMetric);
+                    }
+
+                    // if new metric or metric changed, save to local buffer
+                    this.metrics[scrapedMetric.GetMetricKey()] = scrapedMetric;
+                }
+
+                Log.LogInformation($"Scraped {numScrapedMetrics} Metrics");
+
+                if (metricsToPersist.Count != 0)
+                {
+                    Log.LogInformation("Storing Metrics");
+                    this.storage.WriteData(Newtonsoft.Json.JsonConvert.SerializeObject(metricsToPersist));
+                    Log.LogInformation("Stored Metrics");
+                }
+            }
+        }
+
+        async Task Upload(CancellationToken cancellationToken)
+        {
+            using (await this.scrapeUploadLock.LockAsync(cancellationToken))
+            {
+                Log.LogInformation("Uploading Metrics");
+                await this.uploader.UploadAsync(this.GetMetricsToUpload(this.lastUploadTime), cancellationToken);
+                Log.LogInformation("Uploaded Metrics");
+
+                this.storage.RemoveOldEntries(this.lastUploadTime);
+                this.metrics.Clear();
+                this.lastUploadTime = this.systemTime.UtcNow;
+            }
+        }
+
+        IEnumerable<Metric> GetMetricsToUpload(DateTime lastUploadTime)
```

Comment:
This method is just to get metric based on given timestamp.  Therefore I suggest below change.
```suggestion
        IEnumerable<Metric> GetMetrics(DateTime createdOnOrAfter)
```

Reply:
I think GetMetricsToUpload shows the intent more clearly. This is only ever used by the upload function, so GetMetricsToUpload is exactly what its doing.